src/v/storage/storage_resources.cc

```diff
@@ -91,9 +92,21 @@ void storage_resources::update_allowance(uint64_t total, uint64_t free) {
     _falloc_step = calc_falloc_step();
 }
 
+void storage_resources::update_min_checkpoint_bytes() {
+    if (_partition_count) {
+        // Limit the checkpoint frequency to 2x what it would be under
+        // uniform traffic to all partitions. This permits us to overshoot
+        // target_replay_bytes by approx 50% under non-uniform writes.
```

Comment:
I was thinking about the worst-case bound for this approach, and I think it might be unbounded (though at least only logarithm) in partition count.

E.g., 1 partition uses up the dirty bytes limit, then a second one is added and we use up just under the new threshold of 0.5 * ( DB  / 2 ), then a third partition is added and we use again just under the limit: 0.5 * ( DB / 3 ) ...

So it's harmonic series: 0.5 * (H(n) - 1), which I thing comes out to ~4.4 for 10,000 partitions. So not terrible since I assume the value of 10 GB is kind of arbitrary anyway and 44 GB won't be that much worse. Also this is a very unrealistic write pattern. I do wonder if there is a worse pattern, however.